tests/contract_testcases/solana/expressions/overflow_constant_propagation.sol

```diff
@@ -0,0 +1,62 @@
+// Note: none of these errors will occur without --no-constant-folding
+contract foo {
+    function test() public pure {
+        uint32 x = 2147483648;
+        uint32 y = 2147483648;
+        uint32 z = x + y;
+        print("z: {}".format(z));
+
+        uint16 a1 = 256;
+        uint16 b1 = 257;
+        uint16 c1 = a1 * b1;
+        print("c1: {}".format(c1));
+
+        uint16 a2 = 10;
+        uint16 b2 = 5;
+        uint16 c2 = a2 ** b2;
+        print("c2: {}".format(c2));
+
+        uint16 a3 = 256;
+        uint16 b3 = 257;
+        uint16 c3 = a3 - b3;
+        print("c3: {}".format(c3));
+
+        int16 a4 = -0x8000;
+        int16 c4 = -a4;
+        print("c4: {}".format(c4));
+    }
+
+    function test_unchecked() public pure {
+        unchecked {
+            uint32 x = 2147483648;
+            uint32 y = 2147483648;
+            uint32 z = x + y;
+            print("z: {}".format(z));
+
+            uint16 a1 = 256;
+            uint16 b1 = 257;
+            uint16 c1 = a1 * b1;
+            print("c1: {}".format(c1));
+
+            uint16 a2 = 10;
+            uint16 b2 = 5;
+            uint16 c2 = a2 ** b2;
+            print("c2: {}".format(c2));
+
+            uint16 a3 = 256;
+            uint16 b3 = 257;
+            uint16 c3 = a3 - b3;
+            print("c3: {}".format(c3));
+
+            int16 a4 = -0x8000;
+            int16 c4 = -a4;
+            print("c4: {}".format(c4));
+        }
+    }
+}
+// ---- Expect: diagnostics ----
+// error: 6:20-25: arithmetic overflow: 4294967296 does not fit into uint32
```

Comment:
Are you sure that printing the number is a good idea?
```solidity
contract Test {
        function foo() public pure returns (uint256) {
                uint256 a = 1 << 255;
                uint64 b = 1 << 3;
                return a**b;
        }
}
```

```
error: arithmetic overflow: 126238304966058622268417487065116999845484776053576109500509161826268184136202698801551568013761380717534054534851164138648904527931605160527688095259563605939964364716019515983399209962459578542172100149937763938581219604072733422507180056009672540900709554109516816573779593326332288314873251559077853068444977864803391962580800682760017849589281937637993445539366428356761821065267423102149447628375691862210717202025241630303118559188678304314076943801692528246980959705901641444238894928620825482303431806955690226308773426829503900930529395181208739591967195841536053143145775307050594328881077553168201547776 does not fit into uint256
```

Reply:
Also, trying to calculate can be very slow:
 ```solidity
contract Test {
        function foo() public pure returns (uint256) {
                uint256 a = 1 << 255;
                uint64 b = 1 << 31;
                return a**b;
        }
}
```

Reply:
> Also, trying to calculate can be very slow:
> 
> ```solidity
> contract Test {
>        function foo() public pure returns (uint256) {
>                uint256 a = 1 << 255;
>                uint64 b = 1 << 31;
>                return a**b;
>        }
> }
> ```

This calculation will require too much memory as well.

Reply:
Nice catch. I've re-written this and added these examples to the test

Reply:
I couldn't find the examples